### ISSUE TITLE
DatabaseConnections Metric filter randomly matching active databases.

### ISSUE THREAD

**Issue by wallabyies** [cloud-custodian/cloud-custodian#4014]
I'm using the filters below, and it will periodically (maybe 5% of runs), match on actively used databases.     Any thoughts on this? 

```yaml
filters:
  - type: metrics
    name: DatabaseConnections
    days: 30
    value: 0
    op: equal
    statistics: Sum
  - type: value
    value_type: age
    key: InstanceCreateTime
    value: 29
    op: greater-than
  - "tag:CustodianDeleteThisRDSDatabaseOn": absent 
```
Here is the run log with no changes to the policy, randomly matching on actively used databases:

```
2019-05-13 15:24:46,912 - custodian.policy - INFO - policy: rds-stale-databases resource:rds region:us-east-1 count:14 time:6.01
2019-05-13 15:28:29,274 - custodian.policy - INFO - policy: rds-stale-databases resource:rds region:us-east-1 count:14 time:6.26
2019-05-13 15:31:03,246 - custodian.policy - INFO - policy: rds-stale-databases resource:rds region:us-east-1 count:14 time:6.55
2019-05-13 15:34:24,437 - custodian.policy - INFO - policy: rds-stale-databases resource:rds region:us-east-1 count:14 time:5.85
2019-05-13 15:36:44,139 - custodian.policy - INFO - policy: rds-stale-databases resource:rds region:us-east-1 count:23 time:6.24
```

**Comment by kapilt** [comment]
What version of custodian? This is partial content on the policy, is that all the filters on the original? and the resource here is `rds`?

I just re-reviewed the code here, i'm not seeing anything that would be odd here, its a straight passthrough basically, and if we get no results from the api we'll skip/omit the resource from the filter. I'll see if i can reproduce.

**Comment by wallabyies** [comment]
Thanks!  Here is the version we're using:

c7n:0.8.43.1

Here is the full policy, I'm only running with dryrun from the command line.

```yaml
policies:
  - name: rds-stale-databases
    resource: rds
    filters:
      # Find databases that have had no connections made over the past 30 days.
      - type: metrics
        name: DatabaseConnections
        days: 30
        value: 0
        op: equal
        statistics: Sum
      # Only match on databases >= 30 days old.
      - type: value
        value_type: age
        key: InstanceCreateTime
        value: 29
        op: greater-than
      # Only match on databases that do not have delete date already
      - "tag:CustodianDeleteThisRDSDatabaseOn": absent          
    actions:
       # Tag stale to shutdown.   
       - type: tag
         key: Scheduled
         value: scheduled-stale
       # Tag informative message
       - type: tag
         key: CloudCustodianLastEnforcement
         value: "rds-stale-databases - No connections made in 30 days.   1) Please delete or archive database."        
       # Tag with date to delete, 30 days in future
       - type: tag
         key: CustodianDeleteThisRDSDatabaseOn
         value: "{now:+30d%Y-%m-%d}"
       # Need to confirm behavior of this action, may need to revise.
       - type: mark-for-op
         tag: CustodianDeleteRDSOn
         msg: "Unused RDS Database for >= 60 days: {op}@{action_date}"
         op: notify
         days: 30
```

**Comment by kapilt** [comment]
i haven't been able to reproduce using s3 buckets, @thisisshi are you able to look at this? i don't have many rds instances in my environments anymore.

**Comment by thisisshi** [comment]
I'm also unable to reproduce on my end, like @kapilt said there's nothing in this policy or the metrics filter that looks out of place to me, i've run the policy around 25 times at this point without any changes in resource count

**Comment by wallabyies** [comment]
I'll do some additional testing here on my side and see if I can replicate it again.      I should have an update next week.

Thanks

**Comment by kapilt** [comment]
it might also be helpful to run with -v so we can see the api calls being made per policy run, thanks

**Comment by wallabyies** [comment]
I'm still able to replicate the issue.     I ran with verbose, and output is below.   

As a next step, I will deploy this as a lambda with an action of notify to see if the behavior changes.    All testing up to this point has been command line only.

```
Bad

/custodian$ ./bin/custodian run --dryrun --verbose --cache-period 0 -s . /mnt/c/GitRepo/cloud-custodian-temp/policies/sb/rds-stale-databases.yml
2019-05-18 10:59:39,831: custodian.cache:DEBUG Disabling cache
2019-05-18 10:59:39,832: custodian.commands:DEBUG Loaded file /mnt/c/GitRepo/cloud-custodian-temp/policies/sb/rds-stale-databases.yml. Contains 1 policies
2019-05-18 10:59:39,874: custodian.aws:DEBUG using default region:us-east-1 from boto
2019-05-18 10:59:42,304: custodian.output:DEBUG Storing output with <LogFile file://./rds-stale-databases/custodian-run.log>
2019-05-18 10:59:42,337: custodian.policy:DEBUG Running policy rds-stale-databases resource: rds region:us-east-1 c7n:0.8.43.1
2019-05-18 10:59:46,680: custodian.filters:DEBUG Querying metrics for 38
2019-05-18 10:59:57,647: custodian.resources.rds:DEBUG Filtered from 38 to 25 rds
2019-05-18 10:59:57,650: custodian.policy:INFO policy: rds-stale-databases resource:rds region:us-east-1 count:25 time:15.31
2019-05-18 10:59:57,681: custodian.policy:DEBUG dryrun: skipping actions
2019-05-18 10:59:57,688: custodian.output:DEBUG metric:ResourceCount Count:25 policy:rds-stale-databases restype:rds scope:policy
2019-05-18 10:59:57,691: custodian.output:DEBUG metric:ApiCalls Count:40 policy:rds-stale-databases restype:rds

Good

/custodian$ ./bin/custodian run --dryrun --verbose --cache-period 0 -s . /mnt/c/GitRepo/cloud-custodian-temp/policies/sb/rds-stale-databases.yml
2019-05-18 11:00:11,634: custodian.cache:DEBUG Disabling cache
2019-05-18 11:00:11,636: custodian.commands:DEBUG Loaded file /mnt/c/GitRepo/cloud-custodian-temp/policies/sb/rds-stale-databases.yml. Contains 1 policies
2019-05-18 11:00:11,677: custodian.aws:DEBUG using default region:us-east-1 from boto
2019-05-18 11:00:14,059: custodian.output:DEBUG Storing output with <LogFile file://./rds-stale-databases/custodian-run.log>
2019-05-18 11:00:14,095: custodian.policy:DEBUG Running policy rds-stale-databases resource: rds region:us-east-1 c7n:0.8.43.1
2019-05-18 11:00:19,170: custodian.filters:DEBUG Querying metrics for 38
2019-05-18 11:00:36,058: custodian.resources.rds:DEBUG Filtered from 38 to 13 rds
2019-05-18 11:00:36,061: custodian.policy:INFO policy: rds-stale-databases resource:rds region:us-east-1 count:13 time:21.96
2019-05-18 11:00:36,082: custodian.policy:DEBUG dryrun: skipping actions
2019-05-18 11:00:36,092: custodian.output:DEBUG metric:ResourceCount Count:13 policy:rds-stale-databases restype:rds scope:policy
2019-05-18 11:00:36,097: custodian.output:DEBUG metric:ApiCalls Count:40 policy:rds-stale-databases restype:rds

```

**Comment by wallabyies** [comment]
I deployed this policy as a Lambda, and I'm seeing the same problem/random behavior with active RDS databases being detected by c7n as having no connections in 30 days.

**Comment by wallabyies** [comment]
There may be some related conversation on the Gitter channel.

If you search for "**rds statistics**", find the conversation on July 14, 2017.   One point made was:

**"When I looked at the metrics in resources.json when using "period", I seem to recall it was flagging any resource that went a day without a connection (as you guessed)"**

While, I'm not including a period in my policy, it is similar behavior with rds with non-zero datapoints getting flagged.

**Comment by wallabyies** [comment]
I believe I may have found the cause of this behavior.    In the policy, the metric filter that had issues was:

```
      - type: metrics
        name: DatabaseConnections
        days: 30
        value: 0
        op: equal
        statistics: Sum
```

I changed it to the following.  Note I moved statistics following days.

```
      - type: metrics
        name: DatabaseConnections
        days: 30
        statistics: Sum        
        value: 0
        op: equal
```

In initial testing, this resolved the problem.   I believe if statistics is at the end, it is not being processed, defaulting to "average" and causing the unpredictable results.     

I'll do more thorough testing to confirm this is the case.

**Comment by wallabyies** [comment]
Changing the order of where "statistics" was in the filter did not resolve the problem after all.

**Comment by kapilt** [comment]
is the account under stress? ie are there are lot of things calling the api, just wondering if its a degradation of the cloud watch metrics api. i'm considering if we need to track retries / throttles as an additional metric.

**Comment by wallabyies** [comment]
I've tested this on around five accounts with an average of ~30-40 RDS databases/account.     I was able to replicate the issue on all accounts.

I set the action to notify via email.   When the policy returns false information, the subnet info is not include in the email notification.   When it works fine, sbunet info is is included.   (If that's a clue.)

Reviewing Gitter conversations, there seems to be similar problems if the statistic defaults to average, but I'm specifying Sum in this policy.

**Comment by pendyalal** [comment]
Same, I'm having issue with metrics for RDS DBConnections.

**Comment by ajkerrigan** [comment]
@neilharris123 would you be able to share your custodian version, and a sanitized sample of the logs where you're seeing this behavior?

For context, there was an issue fixed in 0.9.9 (via #6329) that could cause intermittent false positives like the one you're reporting here. If you're still seeing false positives in custodian 0.9.9+, we'll want to sort out the details of what's going wrong.

**Comment by wallabyies** [comment]
I'm closing this out, as I believe this is resolved.     I revisited this in version 0.9.22, and I cannot reproduce the problem.    (I'm also filtering out aurora db engines now which may have helped.)
